### PR TITLE
Keep for QML the better names of list properties

### PR DESCRIPTION
**Pull request by ddobrev**

### CODE REVIEW

**Comment by semlanik** [comment]
Hi, thanks for your contribution! But I could not accept this change. I would like to keep the consistency of .proto properties and Qt properties. Qml property is only an alias for easier access to the QList data. So it has an inappropriate name, probably it should be suffixed as QmlAlias or smt, but the name of Qt property should remain the same as the name of .proto property. It's especially sensitive for Json serialization.

**Comment by ddobrev** [comment]
@semlanik I think what you mean in this case is consistency for the sake of consistency rather than actual usefulness. The QList properties are never actually used because they're inaccessible in `QML` - its constituent members are used instead, as always in `C++`-only usage, and I can see I've kept the nice names for them. That is, the `QQmlListProperty` isn't just an alias, it actually makes the property work in `QML`. And you're suggesting to make the name even worse, just look at this piece of `QML`: `value: object.listQmlAlias[0]`.
About the `JSON` serialization, I realize the build for my PR has failed but as you can see, it was extremely late last night (EET) and I decided to spare myself the effort until I know if you like the idea or not. Please reconsider, and I'll fix all tests so that you can merge.

**Comment by semlanik** [comment]
@ddobrev, yeah your statement partially correct. But test failed not because tests are broken, but because JSON serializing is broken. So in JSON keys should be consistent with the names of a .proto property. I already did some solution for _p AFAIR, probably this would be suitable in this case too. I will appreciate it if you could check this.

**Comment by ddobrev** [comment]
@semlanik I'm extremely glad we've started to reach an understanding. :) I'll be happy to help you so could you please point me to the code you're working on?

**Comment by semlanik** [comment]
https://github.com/semlanik/qtprotobuf/blob/master/src/protobuf/qprotobufmetaproperty.cpp#L51
Yep, this function does some tricks for _p and _proto stuff. But in your case, it wouldn't be so easy, since you couldn't determine at runtime if the property is QList<> or not. I'm thinking about increasing the complexity of the property ordering map and collect the list of aliases in the appropriate structure. Like:
std::unordered_map<int/*proto property*/, struct{int qtProperty, int jsonAlias}>
Need to dig in this direction, but the solution looks promising for me.

**Comment by ddobrev** [comment]
Thank you, I have a better idea now. I'll let you know when it's ready.

**Comment by ddobrev** [comment]
I've added these declarations:

```
constexpr const char *listSuffix = "Data";
constexpr int listSuffixLength = constexpr_strlen(listSuffix);
```

I'm trying this code at present:
```
    if (protoName.endsWith(listSuffix) && QString(typeName()).startsWith("QList<QSharedPointer<")) {
        return protoName.mid(0, protoName.size() - listSuffixLength);
    }
```

If it doesn't work, I'm thinking about this:

```
    if (protoName.endsWith(listSuffix) && QString(typeName()).endsWith("Repeated")) {
        return protoName.mid(0, protoName.size() - listSuffixLength);
    }
```

Please let me know if I'm on the right path.

**Comment by semlanik** [comment]
@ddobrev as far as I remember repeated types are registered using their type aliases names, which contains the "Repeated" suffix, so the basic approach is correct. But I'm afraid there is an issue with this solution, that you probably should check:
What if a user message name is "MyVeryNiceMessage**Repeated**". In this case, I 100% sure, this check will faulty positive. Also extending the protoPropertyName function slows down JSON serialization. This we may probably ignore.

**Comment by ddobrev** [comment]
My second attempt failed too. The other major issue I have is that tests fail locally for me with a clean master, never mind my branch. It looks like sth's wrong with my configuration even though everything built and examples run. What do you think this might be?

**Comment by semlanik** [comment]
> My second attempt failed too. The other major issue I have is that tests fail locally for me with a clean master, never mind my branch. It looks like sth's wrong with my configuration even though everything built and examples run. What do you think this might be?

You need to modify tests that check propertyName() of list properties. Currently, it checks for propertyName without Data suffix.

**Comment by ddobrev** [comment]
I've fixed the tests, let's see...

**Comment by ddobrev** [comment]
@semlanik the build's passed! Is this ready to merge now?

File: tests/test_protobuf/jsondeserializationtest.cpp [hunk 1]
```diff
@@ -319,172 +319,172 @@ TEST_F(JsonDeserializationTest, ComplexTypeSerializeTest)
 TEST_F(JsonDeserializationTest, RepeatedIntMessageTest)
 {
     RepeatedIntMessage test;
-    test.deserialize(serializer.get(), "{\"testRepeatedInt\":[0,1,321,-65999,123245,-3,3]}");
+    test.deserialize(serializer.get(), "{\"testRepeatedIntData\":[0,1,321,-65999,123245,-3,3]}");
```
Comment (semlanik) [inline tests/test_protobuf/jsondeserializationtest.cpp:322]: as long as you change serializer tests, that causes the broken compatibility with the original protobuf.

**Review by semlanik — CHANGES_REQUESTED** [review]

File: tests/test_protobuf/jsondeserializationtest.cpp (hunk 1 above)
Comment (ddobrev) [reply tests/test_protobuf/jsondeserializationtest.cpp:322]: I want to help further but this time I don't understand what to do. If I revert this change, the test's going to fail. Please help me with more instructions.

Comment (ddobrev) [reply tests/test_protobuf/jsondeserializationtest.cpp:322]: @semlanik hello? Please help - I need to know what you want changed.

Comment (semlanik) [reply tests/test_protobuf/jsondeserializationtest.cpp:322]: Let me check your commit I'll back with fix soon :)

Comment (ddobrev) [reply tests/test_protobuf/jsondeserializationtest.cpp:322]: Thank you for looking into this. :) By the way, I got qtprotobuf working with my real project last night. I have some questions which I cannot tell for certain are bugs yet so what's the best way to send them to you?

Comment (semlanik) [reply tests/test_protobuf/jsondeserializationtest.cpp:322]: You always may submit an issue to the project, I don't mind the bugs, etc. at all. I will determine how to deal with such kinds of questions. Eventually, there is the "question" category for issues, and I'm trying to answer ASAP.

File: src/protobuf/qprotobufmetaproperty.cpp [hunk 2]
```diff
@@ -37,8 +37,10 @@ int constexpr constexpr_strlen(const char* str)
 const std::vector<std::string> ListOfQmlExeptions{"id", "property", "import"};
 constexpr const char *privateSuffix = "_p";
 constexpr const char *protoSuffix = "_proto";
-constexpr int privateSuffixLenght = constexpr_strlen(privateSuffix);
-constexpr int protoSuffixLenght = constexpr_strlen(protoSuffix);
+constexpr const char *listSuffix = "Data";
```
Comment (semlanik) [inline src/protobuf/qprotobufmetaproperty.cpp:40]: Please revert this change

File: src/protobuf/qprotobufmetaproperty.cpp [hunk 3]
```diff
@@ -55,15 +57,19 @@ QString QProtobufMetaProperty::protoPropertyName() const
             || userType() == qMetaTypeId<QtProtobuf::fixed32>()
             || userType() == qMetaTypeId<QtProtobuf::sfixed32>())
             && protoName.endsWith(privateSuffix)) {
-        return protoName.mid(0, protoName.size() - privateSuffixLenght);
+        return protoName.mid(0, protoName.size() - privateSuffixLength);
```
Comment (semlanik) [inline src/protobuf/qprotobufmetaproperty.cpp:60]: Please revert this change

File: src/protobuf/qprotobufmetaproperty.cpp [hunk 4]
```diff
@@ -55,15 +57,19 @@ QString QProtobufMetaProperty::protoPropertyName() const
...
     }
 
     if (protoName.endsWith(protoSuffix)) {
-        auto tmpProtoName = protoName.mid(0, protoName.size() - protoSuffixLenght);
+        auto tmpProtoName = protoName.mid(0, protoName.size() - protoSuffixLength);
         if (std::find(ListOfQmlExeptions.begin(), ListOfQmlExeptions.end(), protoName.toStdString()) != ListOfQmlExeptions.end()) {
             return tmpProtoName;
         }
     }
+
```
Comment (semlanik) [inline src/protobuf/qprotobufmetaproperty.cpp:69]: Please revert this change

File: tests/test_protobuf/jsondeserializationtest.cpp (hunk 1 above)
Comment (semlanik) [inline tests/test_protobuf/jsondeserializationtest.cpp:322]: Please revert all changes in tests

**Review by semlanik — CHANGES_REQUESTED** [review]
@ddobrev please undo all changes except the one that actually changes the name of the property. I will forcefully merge this pull request and then post the appropriate fix. Thanks, and sorry for the inconvinience.

**Comment by semlanik** [comment]
Thanks, I will submit fixes a bit later.

**Comment by ddobrev** [comment]
@semlanik thank you for dealing with this. I apologize for the delay as I was asleep earlier.
